ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,222 @@
+import java.util.List;
+import java.util.Objects;
+
+import io.vertx.core.Vertx;
+
+class RecursivePeerRefreshState {
+  private final int CONCURRENT_REQUEST_LIMIT = 3;
+  private final int TIMEOUT_TASK_DELAY = 30000; // 30 Seconds
+  private final BytesValue target;
+  private final PeerBlacklist peerBlacklist;
+  private final BondingAgent bondingAgent;
+  private final NeighborFinder neighborFinder;
+  private final List<PeerDistance> anteList;
+  private final List<OutstandingRequest> outstandingRequestList;
+  private final List<BytesValue> contactedInCurrentExecution;
+  private final Vertx vertx;
+
+  RecursivePeerRefreshState(
+      final BytesValue target,
+      final PeerBlacklist peerBlacklist,
+      final BondingAgent bondingAgent,
+      final NeighborFinder neighborFinder,
+      final Vertx vertx) {
+    this.target = target;
+    this.peerBlacklist = peerBlacklist;
+    this.bondingAgent = bondingAgent;
+    this.neighborFinder = neighborFinder;
+    this.vertx = vertx;
+    this.anteList = new ArrayList<>();
+    this.outstandingRequestList = new ArrayList<>();
+    this.contactedInCurrentExecution = new ArrayList<>();
+
+    commenceTimeoutTask();
+  }
+
+  private void commenceTimeoutTask() {
+    vertx.setPeriodic(
+        TIMEOUT_TASK_DELAY,
+        v -> {
+          List<OutstandingRequest> outstandingRequestListCopy =
+              new ArrayList<>(outstandingRequestList);
+
+          for (OutstandingRequest outstandingRequest : outstandingRequestListCopy) {
+            if (outstandingRequest.isExpired()) {
+              List<Peer> queryCandidates = determineFindNodeCandidates(anteList.size());
+              for (Peer candidate : queryCandidates) {
+                if (!contactedInCurrentExecution.contains(candidate.getId())
+                    && !outstandingRequestList.contains(new OutstandingRequest(candidate))) {
+                  executeFindNodeRequest(candidate);
+                }
+              }
+              outstandingRequestList.remove(outstandingRequest);
+            }
+          }
+        });
+  }
+
+  private void executeFindNodeRequest(final Peer peer) {
+    BytesValue peerId = peer.getId();
+    outstandingRequestList.add(new OutstandingRequest(peer));
+    contactedInCurrentExecution.add(peerId);
+    neighborFinder.issueFindNodeRequest(peer);
+  }
+
+  /**
+   * The lookup initiator starts by picking CONCURRENT_REQUEST_LIMIT closest nodes to the target it
+   * knows of. The initiator then issues concurrent FindNode packets to those nodes.
+   */
+  private void initiatePeerRefreshCycle(final List<Peer> peers) {
+    for (Peer peer : peers) {
+      if (!contactedInCurrentExecution.contains(peer.getId())) {
+        executeFindNodeRequest(peer);
+      }
+      // The lookup terminates when the initiator has queried
+      // and gotten responses from the k closest nodes it has seen.
+    }
+  }
+
+  void digestNeighboursPacket(final NeighborsPacketData neighboursPacket, final Peer peer) {
```

Comment:
I'd suggest moving this and `kickstratBootstrapPeers` up to right under the constructor since they're the key entry point to the class. It's a minor thing but makes it easier to see how the class is used.

I'd probably also call this `onNeighboursPacketReceived` to be a bit more idiomatic.